src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs

```diff
@@ -113,6 +110,94 @@ internal partial class CSharpIntroduceVariableService
             return document.Document.WithSyntaxRoot(newRoot);
         }
 
+        private static bool ShouldIncludeReturnStatement(
+            SemanticDocument document,
+            LambdaExpressionSyntax oldLambda,
+            CancellationToken cancellationToken)
+        {
+            if (document.SemanticModel.GetTypeInfo(oldLambda, cancellationToken).ConvertedType is INamedTypeSymbol delegateType &&
+                delegateType.DelegateInvokeMethod != null)
+            {
+                if (delegateType.DelegateInvokeMethod.ReturnsVoid)
+                {
+                    return false;
+                }
+
+                // Async lambdas with a Task or ValueTask return type don't need a return statement.
+                // e.g.:
+                //     Func<int, Task> f = async x => await M2();
+                //
+                // After refactoring:
+                //     Func<int, Task> f = async x =>
+                //     {
+                //         Task task = M2();
+                //         await task;
+                //     };
+                var compilation = document.SemanticModel.Compilation;
+                if (oldLambda.AsyncKeyword != default &&
+                    (compilation.TaskType() != null && compilation.TaskType().Equals(delegateType.DelegateInvokeMethod.ReturnType) ||
+                    compilation.ValueTaskOfType() != null && compilation.ValueTaskOfType().Equals(delegateType.DelegateInvokeMethod.ReturnType)))
+                {
+                    return false;
+                }
```

Comment:
```suggestion
                if (oldLambda.AsyncKeyword != default)
                {
                    if ((compilation.TaskType() != null && compilation.TaskType().Equals(delegateType.DelegateInvokeMethod.ReturnType) ||
                    (compilation.ValueTaskOfType() != null && compilation.ValueTaskOfType().Equals(delegateType.DelegateInvokeMethod.ReturnType)))
                {
                    return false;
                }
```

Reply:
Can we  assume that the ```compilation.TaskType()``` and ```compilation.ValueTaskType()``` calls will never return null? It seems like they both return nullable types. It also doesn't look like ValueTask existed before C# 7.

Reply:
> Can we assume that the compilation.TaskType() and compilation.ValueTaskType() calls will never return null? 

No we cannot.  First, there's no guarantee that hte user will be compiling against dlls that have that. Second, during project load these dlls can be missing (because the PS hasn't added them yet), and so these methods will return 'null' :)  

It's a pita, but it does happen, so it's good to write the code to be resilient to that.  We alst test for this.  for example if you look at the UseRange (or maybe UseIndex) tests, we tests for cases where this stuff runs in projects where the extra dlls are not there.

Reply:
See UseRangeOperationTests.TestWithMissingReference.  This is actually a reasonable idea to test almost all features in this condition as it accurately reflects what might happen during solution load.  No biggie if you do/don't add the test.

Reply:
Gotcha! Will definitely keep that in mind for the future.